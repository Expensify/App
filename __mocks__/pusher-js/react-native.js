import {PusherMock} from 'pusher-js-mock';

class PusherMockWithDisconnect extends PusherMock {
    disconnect() {
        return jest.fn();
    }
}

export default PusherMockWithDisconnect;
