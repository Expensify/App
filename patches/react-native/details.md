# `react-native` patches

### [react-native+0.79.2+001+initial.patch](react-native+0.79.2+001+initial.patch)

- Reason: Fixes keyboard flickering issue when opening/closing modals. When an input is blurred and a modal is opened, the `rootView` becomes the `firstResponder`, causing the system to retain an incorrect keyboard state. This leads to keyboard flickering when the modal is closed. The patch adds code to resign the `rootView`'s `firstResponder` status before presenting the modal to prevent this issue.
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: https://github.com/Expensify/App/pull/23994

### [react-native+0.79.2+002+fixMVCPAndroid.patch](react-native+0.79.2+002+fixMVCPAndroid.patch)

- Reason: Fixes content jumping issues with `MaintainVisibleContentPosition` on Android, particularly in bidirectional pagination scenarios. The patch makes two key improvements:
  1. Changes when the first visible view is calculated - now happens on scroll events instead of during Fabric's willMountItems lifecycle, which was causing incorrect updates
  2. Improves first visible view selection logic to handle Fabric's z-index-based view reordering by finding the view with the smallest position that's still greater than the scroll position
- Upstream PR/issue: https://github.com/facebook/react-native/pull/46247
- E/App issue: ðŸ›‘
- PR Introducing Patch: https://github.com/Expensify/App/pull/46315 (introduced), https://github.com/Expensify/App/pull/45289 (refactored)

### [react-native+0.79.2+003+disableTextInputRecycling.patch](react-native+0.79.2+003+disableTextInputRecycling.patch)

- Reason: Disables text input recycling to prevent issues with state of recycled component
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: https://github.com/Expensify/App/pull/13767

### [react-native+0.79.2+004+iOSFontResolution.patch](react-native+0.79.2+004+iOSFontResolution.patch)

- Reason: Fixes font resolution issues on iOS by properly preserving font properties when loading fonts by name. When a font is loaded by its name, the patch now correctly extracts and uses its actual weight and style properties instead of ignoring them.
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: https://github.com/Expensify/App/pull/13767

### [react-native+0.79.2+005+resetAutoresizingOnView.patch](react-native+0.79.2+005+resetAutoresizingOnView.patch)

- Reason: This is a workaround fix for an issue with `UIPageViewController` and Fabric's view recycling system. The problem occurs because pager-view was incorrectly using a Fabric-provided view as its content-view. This is problematic because `UIPageViewController` modifies its content view, and when Fabric later recycles this modified view, it can lead to unexpected layout issues. The patch addresses this by resetting the autoresizing mask on the view to prevent layout corruption when views are recycled. The root cause should be addressed in react-native-pager-view: https://github.com/callstack/react-native-pager-view/issues/819
- Upstream PR/issue: https://github.com/facebook/react-native/issues/42732
- E/App issue: ðŸ›‘
- PR Introducing Patch: https://github.com/Expensify/App/pull/13767

### [react-native+0.79.2+006+disableNonTranslucentStatusBar.patch](react-native+0.79.2+006+disableNonTranslucentStatusBar.patch)

- Reason: Disables non-translucent status bar to fix UI issues
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+007+TextInput.patch](react-native+0.79.2+007+TextInput.patch)

- Reason: Fixes TextInput component issues
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+008+iOSCoreAnimationBorderRendering.patch](react-native+0.79.2+008+iOSCoreAnimationBorderRendering.patch)

- Reason: Fixes border rendering issues with Core Animation on iOS
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+009+copyStateOnClone.patch](react-native+0.79.2+009+copyStateOnClone.patch)

- Reason: Ensures state is properly copied when cloning components
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+010+textinput-clear-command.patch](react-native+0.79.2+010+textinput-clear-command.patch)

- Reason: Adds clear command functionality to TextInput
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+011+Add-onPaste-to-TextInput.patch](react-native+0.79.2+011+Add-onPaste-to-TextInput.patch)

- Reason: Adds onPaste event handler to TextInput component
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+012+alert-style.patch](react-native+0.79.2+012+alert-style.patch)

- Reason: Fixes alert styling issues
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+013+fixNavigationAnimations.patch](react-native+0.79.2+013+fixNavigationAnimations.patch)

- Reason: Fixes navigation animation issues
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+014+fixScrollViewState.patch](react-native+0.79.2+014+fixScrollViewState.patch)

- Reason: Fixes ScrollView state management issues
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+015+redactAppParameters.patch](react-native+0.79.2+015+redactAppParameters.patch)

- Reason: Adds redaction of sensitive app parameters
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+016+android-keyboard-avoiding-view.patch](react-native+0.79.2+016+android-keyboard-avoiding-view.patch)

- Reason: Fixes keyboard avoiding view behavior on Android
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+017+fix-mask-persisting-recycling.patch](react-native+0.79.2+017+fix-mask-persisting-recycling.patch)

- Reason: Fixes mask persisting and recycling issues
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+018+fix-text-selecting-on-change.patch](react-native+0.79.2+018+fix-text-selecting-on-change.patch)

- Reason: Fixes text selection issues during onChange events
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+019+fix-scroll-view-jump.patch](react-native+0.79.2+019+fix-scroll-view-jump.patch)

- Reason: Fixes ScrollView jumping issues
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+020+fix-dropping-mutations-in-transactions.patch](react-native+0.79.2+020+fix-dropping-mutations-in-transactions.patch)

- Reason: Fixes issues with dropping mutations in transactions
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+021+fix-crash-when-deleting-expense.patch](react-native+0.79.2+021+fix-crash-when-deleting-expense.patch)

- Reason: Fixes crash when deleting expenses
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+022+fix-surface-stopped-before-started.patch](react-native+0.79.2+022+fix-surface-stopped-before-started.patch)

- Reason: Fixes surface lifecycle issues where surface is stopped before being started
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: ðŸ›‘

### [react-native+0.79.2+023+publish-gradle.patch](react-native+0.79.2+023+publish-gradle.patch)

- Reason: This patch customizes the Gradle publishing script to allow publishing our custom React Native artifacts to GitHub Packages.
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: https://github.com/Expensify/App/pull/59738

### [react-native+0.79.2+024+restore-old-line-height-algorithm.patch](react-native+0.79.2+024+restore-old-line-height-algorithm.patch)

- Reason: This patch fixes line height calculation issues in Android text rendering by replacing the web-based CSS implementation with a priority-based approach that properly handles cases where font metrics exceed the specified line height, ensuring better text display consistency and preventing text clipping
- Upstream PR/issue: ðŸ›‘
- E/App issue: ðŸ›‘
- PR Introducing Patch: https://github.com/Expensify/App/pull/60421
