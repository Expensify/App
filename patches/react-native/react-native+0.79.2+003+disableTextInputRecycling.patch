diff --git a/node_modules/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm b/node_modules/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm
index e74500f..c1d7c24 100644
--- a/node_modules/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm
+++ b/node_modules/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm
@@ -155,6 +155,11 @@ - (NSObject *)accessibilityElement
 
 #pragma mark - RCTComponentViewProtocol
 
++ (BOOL)shouldBeRecycled
+{
+    return NO;
+}
+
 + (ComponentDescriptorProvider)componentDescriptorProvider
 {
   return concreteComponentDescriptorProvider<TextInputComponentDescriptor>();
