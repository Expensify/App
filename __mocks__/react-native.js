// eslint-disable-next-line no-restricted-imports
import * as ReactNative from 'react-native';
import _ from 'underscore';
import CONST from '../src/CONST';

jest.doMock('react-native', () => {
    let url = 'https://new.expensify.com/';
    const getInitialURL = () => Promise.resolve(url);

    let appState = 'active';
    let count = 0;
    const changeListeners = {};

    // Tests will run with the app in a typical small screen size by default. We do this since the react-native test renderer
    // runs against index.native.js source and so anything that is testing a component reliant on withWindowDimensions()
    // would be most commonly assumed to be on a mobile phone vs. a tablet or desktop style view. This behavior can be
    // overridden by explicitly setting the dimensions inside a test via Dimensions.set()
    let dimensions = CONST.TESTING.SCREEN_SIZE.SMALL;

    return Object.setPrototypeOf(
        {
            NativeModules: {
                ...ReactNative.NativeModules,
                BootSplash: {
                    getVisibilityStatus: jest.fn(),
                    hide: jest.fn(),
                },
                StartupTimer: {stop: jest.fn()},
            },
            Linking: {
                ...ReactNative.Linking,
                getInitialURL,
                setInitialURL(newUrl) {
                    url = newUrl;
                },
            },
            AppState: {
                ...ReactNative.AppState,
                get currentState() {
                    return appState;
                },
                emitCurrentTestState(state) {
                    appState = state;
                    _.each(changeListeners, listener => listener(appState));
                },
                addEventListener(type, listener) {
                    if (type === 'change') {
                        const originalCount = count;
                        changeListeners[originalCount] = listener;
                        ++count;
                        return {
                            remove: () => {
                                delete changeListeners[originalCount];
                            },
                        };
                    }

                    return ReactNative.AppState.addEventListener(type, listener);
                },
            },
            Dimensions: {
                ...ReactNative.Dimensions,
                addEventListener: jest.fn(),
                get: () => dimensions,
                set: (newDimensions) => {
                    dimensions = newDimensions;
                },
            },
        },
        ReactNative,
    );
});
