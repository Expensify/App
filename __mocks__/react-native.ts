// eslint-disable-next-line no-restricted-imports
import * as ReactNative from 'react-native';
import type StartupTimer from '@libs/StartupTimer/types';

const {BootSplash} = ReactNative.NativeModules;

jest.doMock('react-native', () => {
    let url = 'https://new.expensify.com/';
    const getInitialURL = () => Promise.resolve(url);

    let appState: ReactNative.AppStateStatus = 'active';
    let count = 0;
    const changeListeners: Record<number, (state: ReactNative.AppStateStatus) => void> = {};

    // Tests will run with the app in a typical small screen size by default. We do this since the react-native test renderer
    // runs against index.native.js source and so anything that is testing a component reliant on withWindowDimensions()
    // would be most commonly assumed to be on a mobile phone vs. a tablet or desktop style view. This behavior can be
    // overridden by explicitly setting the dimensions inside a test via Dimensions.set()
    let dimensions: Record<string, number> = {
        width: 300,
        height: 700,
        scale: 1,
        fontScale: 1,
    };

    type ReactNativeMock = typeof ReactNative & {
        NativeModules: typeof ReactNative.NativeModules & {
            BootSplash: {
                getVisibilityStatus: typeof BootSplash.getVisibilityStatus;
                hide: typeof BootSplash.hide;
                logoSizeRatio: number;
                navigationBarHeight: number;
            };
            StartupTimer: StartupTimer;
        };
        Linking: typeof ReactNative.Linking & {
            setInitialURL: (newUrl: string) => void;
        };
        AppState: typeof ReactNative.AppState & {
            emitCurrentTestState: (state: ReactNative.AppStateStatus) => void;
        };
    };

    const reactNativeMock: ReactNativeMock = Object.setPrototypeOf(
        {
            NativeModules: {
                ...ReactNative.NativeModules,
                BootSplash: {
                    getVisibilityStatus: jest.fn(),
                    hide: jest.fn(),
                    logoSizeRatio: 1,
                    navigationBarHeight: 0,
                },
                StartupTimer: {stop: jest.fn()},
            },
            Linking: {
                ...ReactNative.Linking,
                getInitialURL,
                setInitialURL(newUrl: string) {
                    url = newUrl;
                },
            },
            AppState: {
                ...ReactNative.AppState,
                get currentState() {
                    return appState;
                },
                emitCurrentTestState(state: ReactNative.AppStateStatus) {
                    appState = state;
                    Object.entries(changeListeners).forEach(([, listener]) => listener(appState));
                },
                addEventListener(type: ReactNative.AppStateEvent, listener: (state: ReactNative.AppStateStatus) => void) {
                    if (type === 'change') {
                        const originalCount = count;
                        changeListeners[originalCount] = listener;
                        ++count;
                        return {
                            remove: () => {
                                delete changeListeners[originalCount];
                            },
                        };
                    }

                    return ReactNative.AppState.addEventListener(type, listener);
                },
            },
            Dimensions: {
                ...ReactNative.Dimensions,
                addEventListener: jest.fn(),
                get: () => dimensions,
                set: (newDimensions: Record<string, number>) => {
                    dimensions = newDimensions;
                },
            },

            // `runAfterInteractions` method would normally be triggered after the native animation is completed,
            // we would have to mock waiting for the animation end and more state changes,
            // so it seems easier to just run the callback immediately in tests.
            InteractionManager: {
                ...ReactNative.InteractionManager,
                runAfterInteractions: (callback: () => void) => callback(),
            },
        },
        ReactNative,
    );

    return reactNativeMock;
});
