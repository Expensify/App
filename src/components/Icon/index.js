import React, {PureComponent} from 'react';
import {View} from 'react-native';
import PropTypes from 'prop-types';
import styles from '../../styles/styles';
import themeColors from '../../styles/themes/default';
import variables from '../../styles/variables';
import * as StyleUtils from '../../styles/StyleUtils';
import IconWrapperStyles from './IconWrapperStyles';

const propTypes = {
    /** The asset to render. */
    src: PropTypes.func.isRequired,

    /** The width of the icon. */
    width: PropTypes.number,

    /** The height of the icon. */
    height: PropTypes.number,

    /** The fill color for the icon. Can be hex, rgb, rgba, or valid react-native named color such as 'red' or 'blue' */
    fill: PropTypes.string,

    /** Is small icon */
    small: PropTypes.bool,

    /** Is inline icon */
    inline: PropTypes.bool,
};

const defaultProps = {
    width: variables.iconSizeNormal,
    height: variables.iconSizeNormal,
    fill: themeColors.icon,
    small: false,
    inline: false,
};

// We must use a class component to create an animatable component with the Animated API
// eslint-disable-next-line react/prefer-stateless-function
class Icon extends PureComponent {
    render() {
        const width = this.props.small ? variables.iconSizeSmall : this.props.width;
        const height = this.props.small ? variables.iconSizeSmall : this.props.height;

        if (this.props.inline) {
            return (
                <View
                    accessibilityHint={`${this.props.src.name} Icon`}
                    style={[StyleUtils.getWidthAndHeightStyle(width, height), styles.bgTransparent, styles.overflowVisible]}
                >
                    <View style={[StyleUtils.getWidthAndHeightStyle(width, height), IconWrapperStyles, styles.pAbsolute]}>
                        <this.props.src
                            width={width}
                            height={height}
                            fill={this.props.fill}
                        />
                    </View>
                </View>
            );
        }

        return (
            <View accessibilityHint={`${this.props.src.name} Icon`}>
                <this.props.src
                    width={width}
                    height={height}
                    fill={this.props.fill}
                />
            </View>
        );
    }
}

Icon.propTypes = propTypes;
Icon.defaultProps = defaultProps;

export default Icon;
