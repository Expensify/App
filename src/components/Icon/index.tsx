import React, {PureComponent} from 'react';
import {StyleProp, View, ViewStyle} from 'react-native';
import styles from '@styles/styles';
import * as StyleUtils from '@styles/StyleUtils';
import themeColors from '@styles/themes/default';
import variables from '@styles/variables';
import IconWrapperStyles from './IconWrapperStyles';

type SrcProps = {
    width?: number;
    height?: number;
    fill?: string;
    hovered?: string;
    pressed?: string;
};

type IconProps = {
    /** The asset to render. */
    src: (props: SrcProps) => React.ReactNode;

    /** The width of the icon. */
    width?: number;

    /** The height of the icon. */
    height?: number;

    /** The fill color for the icon. Can be hex, rgb, rgba, or valid react-native named color such as 'red' or 'blue'. */
    fill?: string;

    /** Is small icon */
    small?: boolean;

    /** Is inline icon */
    inline?: boolean;

    /** Is icon hovered */
    hovered?: boolean;

    /** Is icon pressed */
    pressed?: boolean;

    /** Additional styles to add to the Icon */
    additionalStyles?: StyleProp<ViewStyle>;
};

// We must use a class component to create an animatable component with the Animated API
// eslint-disable-next-line react/prefer-stateless-function
class Icon extends PureComponent<IconProps> {
    // eslint-disable-next-line react/static-property-placement
    public static defaultProps = {
        width: variables.iconSizeNormal,
        height: variables.iconSizeNormal,
        fill: themeColors.icon,
        small: false,
        inline: false,
        additionalStyles: [],
        hovered: false,
        pressed: false,
    };

    render() {
        const width = this.props.small ? variables.iconSizeSmall : this.props.width;
        const height = this.props.small ? variables.iconSizeSmall : this.props.height;
        const iconStyles = [StyleUtils.getWidthAndHeightStyle(width ?? 0, height), IconWrapperStyles, styles.pAbsolute, this.props.additionalStyles];

        if (this.props.inline) {
            return (
                <View
                    testID={`${this.props.src.name} Icon`}
                    style={[StyleUtils.getWidthAndHeightStyle(width ?? 0, height), styles.bgTransparent, styles.overflowVisible]}
                >
                    <View style={iconStyles}>
                        <this.props.src
                            width={width}
                            height={height}
                            fill={this.props.fill}
                            hovered={this.props.hovered?.toString()}
                            pressed={this.props.pressed?.toString()}
                        />
                    </View>
                </View>
            );
        }

        return (
            <View
                testID={`${this.props.src.name} Icon`}
                style={this.props.additionalStyles}
            >
                <this.props.src
                    width={width}
                    height={height}
                    fill={this.props.fill}
                    hovered={this.props.hovered?.toString()}
                    pressed={this.props.pressed?.toString()}
                />
            </View>
        );
    }
}

export default Icon;
