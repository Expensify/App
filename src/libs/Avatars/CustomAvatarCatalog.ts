/* eslint-disable @typescript-eslint/naming-convention */
import type {SvgProps} from 'react-native-svg';
import * as SeasonF1 from '@components/Icon/CustomAvatars/SeasonF1';
import * as DefaultAvatars from '@components/Icon/DefaultAvatars';
import CONST from '@src/CONST';

type DefaultAvatarIDs =
    | 'default-avatar_1'
    | 'default-avatar_2'
    | 'default-avatar_3'
    | 'default-avatar_4'
    | 'default-avatar_5'
    | 'default-avatar_6'
    | 'default-avatar_7'
    | 'default-avatar_8'
    | 'default-avatar_9'
    | 'default-avatar_10'
    | 'default-avatar_11'
    | 'default-avatar_12'
    | 'default-avatar_13'
    | 'default-avatar_14'
    | 'default-avatar_15'
    | 'default-avatar_16'
    | 'default-avatar_17'
    | 'default-avatar_18'
    | 'default-avatar_19'
    | 'default-avatar_20'
    | 'default-avatar_21'
    | 'default-avatar_22'
    | 'default-avatar_23'
    | 'default-avatar_24';

type SeasonF1AvatarIDs =
    | 'car-blue100'
    | 'car-green100'
    | 'car-ice100'
    | 'car-pink100'
    | 'car-tangerine100'
    | 'car-yellow100'
    | 'champagne-green400'
    | 'cone-tangerine700'
    | 'flag-blue600'
    | 'gasoline-tangerine400'
    | 'helmet-blue400'
    | 'helmet-green400'
    | 'helmet-ice400'
    | 'helmet-pink400'
    | 'helmet-tangerine400'
    | 'helmet-yellow400'
    | 'medal-yellow400'
    | 'podium-blue400'
    | 'speedometer-ice400'
    | 'steeringwheel-pink400'
    | 'stopwatch-ice600'
    | 'tire-green400'
    | 'trophy-yellow600'
    | 'wrenches-pink600';

const CDN_DEFAULT_AVATARS = `${CONST.CLOUDFRONT_URL}/images/avatars`;
const CDN_SEASON_F1 = `${CONST.CLOUDFRONT_URL}/images/avatars/custom-avatars/season-f1`;

type AvatarEntry = {local: React.FC<SvgProps>; url: string};
type CustomAvatarID = DefaultAvatarIDs | SeasonF1AvatarIDs;

const DEFAULTS: Record<DefaultAvatarIDs, AvatarEntry> = {
    'default-avatar_1': {local: DefaultAvatars.Avatar1, url: `${CDN_DEFAULT_AVATARS}/default-avatar_1.png`},
    'default-avatar_2': {local: DefaultAvatars.Avatar2, url: `${CDN_DEFAULT_AVATARS}/default-avatar_2.png`},
    'default-avatar_3': {local: DefaultAvatars.Avatar3, url: `${CDN_DEFAULT_AVATARS}/default-avatar_3.png`},
    'default-avatar_4': {local: DefaultAvatars.Avatar4, url: `${CDN_DEFAULT_AVATARS}/default-avatar_4.png`},
    'default-avatar_5': {local: DefaultAvatars.Avatar5, url: `${CDN_DEFAULT_AVATARS}/default-avatar_5.png`},
    'default-avatar_6': {local: DefaultAvatars.Avatar6, url: `${CDN_DEFAULT_AVATARS}/default-avatar_6.png`},
    'default-avatar_7': {local: DefaultAvatars.Avatar7, url: `${CDN_DEFAULT_AVATARS}/default-avatar_7.png`},
    'default-avatar_8': {local: DefaultAvatars.Avatar8, url: `${CDN_DEFAULT_AVATARS}/default-avatar_8.png`},
    'default-avatar_9': {local: DefaultAvatars.Avatar9, url: `${CDN_DEFAULT_AVATARS}/default-avatar_9.png`},
    'default-avatar_10': {local: DefaultAvatars.Avatar10, url: `${CDN_DEFAULT_AVATARS}/default-avatar_10.png`},
    'default-avatar_11': {local: DefaultAvatars.Avatar11, url: `${CDN_DEFAULT_AVATARS}/default-avatar_11.png`},
    'default-avatar_12': {local: DefaultAvatars.Avatar12, url: `${CDN_DEFAULT_AVATARS}/default-avatar_12.png`},
    'default-avatar_13': {local: DefaultAvatars.Avatar13, url: `${CDN_DEFAULT_AVATARS}/default-avatar_13.png`},
    'default-avatar_14': {local: DefaultAvatars.Avatar14, url: `${CDN_DEFAULT_AVATARS}/default-avatar_14.png`},
    'default-avatar_15': {local: DefaultAvatars.Avatar15, url: `${CDN_DEFAULT_AVATARS}/default-avatar_15.png`},
    'default-avatar_16': {local: DefaultAvatars.Avatar16, url: `${CDN_DEFAULT_AVATARS}/default-avatar_16.png`},
    'default-avatar_17': {local: DefaultAvatars.Avatar17, url: `${CDN_DEFAULT_AVATARS}/default-avatar_17.png`},
    'default-avatar_18': {local: DefaultAvatars.Avatar18, url: `${CDN_DEFAULT_AVATARS}/default-avatar_18.png`},
    'default-avatar_19': {local: DefaultAvatars.Avatar19, url: `${CDN_DEFAULT_AVATARS}/default-avatar_19.png`},
    'default-avatar_20': {local: DefaultAvatars.Avatar20, url: `${CDN_DEFAULT_AVATARS}/default-avatar_20.png`},
    'default-avatar_21': {local: DefaultAvatars.Avatar21, url: `${CDN_DEFAULT_AVATARS}/default-avatar_21.png`},
    'default-avatar_22': {local: DefaultAvatars.Avatar22, url: `${CDN_DEFAULT_AVATARS}/default-avatar_22.png`},
    'default-avatar_23': {local: DefaultAvatars.Avatar23, url: `${CDN_DEFAULT_AVATARS}/default-avatar_23.png`},
    'default-avatar_24': {local: DefaultAvatars.Avatar24, url: `${CDN_DEFAULT_AVATARS}/default-avatar_24.png`},
};

const SEASON_F1: Record<SeasonF1AvatarIDs, AvatarEntry> = {
    'car-blue100': {local: SeasonF1.CarBlue100, url: `${CDN_SEASON_F1}/car-blue100.png`},
    'car-green100': {local: SeasonF1.CarGreen100, url: `${CDN_SEASON_F1}/car-green100.png`},
    'car-ice100': {local: SeasonF1.CarIce100, url: `${CDN_SEASON_F1}/car-ice100.png`},
    'car-pink100': {local: SeasonF1.CarPink100, url: `${CDN_SEASON_F1}/car-pink100.png`},
    'car-tangerine100': {local: SeasonF1.CarTangerine100, url: `${CDN_SEASON_F1}/car-tangerine100.png`},
    'car-yellow100': {local: SeasonF1.CarYellow100, url: `${CDN_SEASON_F1}/car-yellow100.png`},
    'champagne-green400': {local: SeasonF1.ChampagneGreen400, url: `${CDN_SEASON_F1}/champagne-green400.png`},
    'cone-tangerine700': {local: SeasonF1.ConeTangerine700, url: `${CDN_SEASON_F1}/cone-tangerine700.png`},
    'flag-blue600': {local: SeasonF1.FlagBlue600, url: `${CDN_SEASON_F1}/flag-blue600.png`},
    'gasoline-tangerine400': {local: SeasonF1.GasolineTangerine400, url: `${CDN_SEASON_F1}/gasoline-tangerine400.png`},
    'helmet-blue400': {local: SeasonF1.HelmetBlue400, url: `${CDN_SEASON_F1}/helmet-blue400.png`},
    'helmet-green400': {local: SeasonF1.HelmetGreen400, url: `${CDN_SEASON_F1}/helmet-green400.png`},
    'helmet-ice400': {local: SeasonF1.HelmetIce400, url: `${CDN_SEASON_F1}/helmet-ice400.png`},
    'helmet-pink400': {local: SeasonF1.HelmetPink400, url: `${CDN_SEASON_F1}/helmet-pink400.png`},
    'helmet-tangerine400': {local: SeasonF1.HelmetTangerine400, url: `${CDN_SEASON_F1}/helmet-tangerine400.png`},
    'helmet-yellow400': {local: SeasonF1.HelmetYellow400, url: `${CDN_SEASON_F1}/helmet-yellow400.png`},
    'medal-yellow400': {local: SeasonF1.MedalYellow400, url: `${CDN_SEASON_F1}/medal-yellow400.png`},
    'podium-blue400': {local: SeasonF1.PodiumBlue400, url: `${CDN_SEASON_F1}/podium-blue400.png`},
    'speedometer-ice400': {local: SeasonF1.SpeedometerIce400, url: `${CDN_SEASON_F1}/speedometer-ice400.png`},
    'steeringwheel-pink400': {local: SeasonF1.SteeringWheelPink400, url: `${CDN_SEASON_F1}/steeringwheel-pink400.png`},
    'stopwatch-ice600': {local: SeasonF1.StopwatchIce600, url: `${CDN_SEASON_F1}/stopwatch-ice600.png`},
    'tire-green400': {local: SeasonF1.TireGreen400, url: `${CDN_SEASON_F1}/tire-green400.png`},
    'trophy-yellow600': {local: SeasonF1.TrophyYellow600, url: `${CDN_SEASON_F1}/trophy-yellow600.png`},
    'wrenches-pink600': {local: SeasonF1.WrenchesPink600, url: `${CDN_SEASON_F1}/wrenches-pink600.png`},
};

const ALL_CUSTOM_AVATARS: Record<CustomAvatarID, AvatarEntry> = {
    ...DEFAULTS,
    ...SEASON_F1,
};

const getAvatarLocal = (id: CustomAvatarID) => ALL_CUSTOM_AVATARS[id].local;
const getAvatarURL = (id: CustomAvatarID) => ALL_CUSTOM_AVATARS[id].url;

export {ALL_CUSTOM_AVATARS, getAvatarLocal, getAvatarURL};
export type {DefaultAvatarIDs, SeasonF1AvatarIDs, CustomAvatarID};
