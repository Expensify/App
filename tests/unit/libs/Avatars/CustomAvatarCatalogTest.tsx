import {render} from '@testing-library/react-native';
import React from 'react';
import Icon from '@components/Icon';
import {ALL_CUSTOM_AVATARS, getAvatarLocal, getAvatarURL} from '@libs/Avatars/CustomAvatarCatalog';

const SAMPLE_DEFAULT_ID = 'default-avatar_1';
const SAMPLE_SEASON_ID = 'car-blue100';

describe('CustomAvatarCatalog', () => {
    it('resolves a local component for a default avatar ID', () => {
        const AvatarComponent = getAvatarLocal(SAMPLE_DEFAULT_ID);
        expect(AvatarComponent).toBeDefined();
        expect(typeof AvatarComponent).toBe('function');
    });

    it('resolves a CDN URL for a default avatar ID', () => {
        const avatarUrl = getAvatarURL(SAMPLE_DEFAULT_ID);
        expect(avatarUrl).toContain(`/images/avatars/${SAMPLE_DEFAULT_ID}`);
    });

    it('renders a default avatar component', () => {
        const AvatarComponent = getAvatarLocal(SAMPLE_DEFAULT_ID);
        const {toJSON} = render(<Icon src={AvatarComponent} />);
        expect(toJSON()).toBeTruthy();
    });

    it('matches snapshot for a default avatar', () => {
        const AvatarComponent = getAvatarLocal(SAMPLE_DEFAULT_ID);
        const {toJSON} = render(<Icon src={AvatarComponent} />);
        expect(toJSON()).toMatchSnapshot();
    });

    it('resolves a local component for a seasonal avatar ID', () => {
        const AvatarComponent = getAvatarLocal(SAMPLE_SEASON_ID);
        expect(AvatarComponent).toBeDefined();
        expect(typeof AvatarComponent).toBe('function');
    });

    it('resolves a CDN URL for a seasonal avatar ID', () => {
        const avatarUrl = getAvatarURL(SAMPLE_SEASON_ID);
        expect(avatarUrl).toContain(`/images/avatars/custom-avatars/season-f1/${SAMPLE_SEASON_ID}`);
    });

    it('renders a seasonal avatar component', () => {
        const AvatarComponent = getAvatarLocal(SAMPLE_SEASON_ID);
        const {toJSON} = render(<Icon src={AvatarComponent} />);
        expect(toJSON()).toBeTruthy();
    });

    it('throws or returns undefined for an unknown ID', () => {
        // @ts-expect-error - This is a test for an unknown ID
        expect(() => getAvatarLocal('not-a-real-id')).toThrow();
    });

    it('ALL contains both default and seasonal IDs', () => {
        expect(Object.keys(ALL_CUSTOM_AVATARS)).toEqual(expect.arrayContaining([SAMPLE_DEFAULT_ID, SAMPLE_SEASON_ID]));
    });
});
