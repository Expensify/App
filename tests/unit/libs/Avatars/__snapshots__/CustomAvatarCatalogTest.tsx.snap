// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomAvatarCatalog matches snapshot for a default avatar 1`] = `
<View
  style={[]}
  testID=""
>
  <svg
    data-file-name="SvgDefaultAvatar_1"
    height={20}
    width={20}
  />
</View>
`;
